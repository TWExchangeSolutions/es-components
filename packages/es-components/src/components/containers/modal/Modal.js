import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { noop } from 'lodash';
import BaseModal from 'react-overlays/lib/Modal';

import genId from '../../util/generateAlphaName';
import { useTheme } from '../../util/useTheme';
import Fade from '../../util/Fade';
import { ModalContext } from './ModalContext';
import Header from './ModalHeader';
import Body from './ModalBody';
import Footer from './ModalFooter';

const modalSize = {
  small: '300px',
  medium: '600px',
  large: '900px'
};

const DialogWrapper = styled(BaseModal)`
  height: 100%;
  left: 0;
  position: fixed;
  text-align: center;
  top: 0;
  width: 100%;
  z-index: 1040;
`;

const ModalDialogMedium = styled.div`
  cursor: default;
  display: inline-block;
  top: 2em;

  @media (min-width: ${props => props.theme.screenSize.tablet}) {
    margin: 30px auto;
    width: ${modalSize.medium};
  }
`;

const ModalDialogSmall = styled(ModalDialogMedium)`
  @media (min-width: ${props => props.theme.screenSize.phone}) {
    margin: 30px auto;
    width: ${modalSize.small};
  }
`;

const ModalDialogLarge = styled(ModalDialogMedium)`
  @media (min-width: ${props => props.theme.screenSize.desktop}) {
    width: ${modalSize.large};
  }
`;

const ModalContent = styled.div`
  background-clip: padding-box;
  background-color: ${props => props.theme.colors.white};
  box-shadow: 0 5px 15px rgba(0, 0, 0, 0.5);
  font-size: ${props => props.theme.sizes.baseFontSize};
  outline: 0;
  position: relative;
`;

function getModalBySize(size) {
  switch (size) {
    case 'small':
      return ModalDialogSmall;
    case 'large':
      return ModalDialogLarge;
    default:
      return ModalDialogMedium;
  }
}

function Modal(props) {
  const {
    animation,
    backdrop,
    children,
    escapeExits,
    onEnter,
    onExit,
    onHide,
    show,
    size
  } = props;

  const theme = useTheme();
  const backdropStyle = {
    backgroundColor: theme.colors.black,
    bottom: 0,
    cursor: backdrop === 'static' ? 'auto' : 'pointer',
    left: 0,
    opacity: 0.5,
    position: 'fixed',
    right: 0,
    top: 0,
    zIndex: 'auto'
  };

  const ModalDialog = getModalBySize(size);

  const ariaId = genId();

  return (
    <ModalContext.Provider value={{ onHide, ariaId }}>
      <DialogWrapper
        backdrop={backdrop}
        backdropStyle={backdropStyle}
        keyboard={escapeExits}
        onEnter={onEnter}
        onExit={onExit}
        onHide={onHide}
        show={show}
        transition={animation ? Fade : undefined}
        className="es-modal"
      >
        <ModalDialog
          size={size}
          aria-labelledby={ariaId}
          className="es-modal__dialog"
        >
          <ModalContent className="es-modal__content">{children}</ModalContent>
        </ModalDialog>
      </DialogWrapper>
    </ModalContext.Provider>
  );
}

Modal.propTypes = {
  /** Open and close the Modal with a fade animation. */
  animation: PropTypes.bool,
  /**
   * Include a backdrop component. Specify 'static' for a backdrop that doesn't
   * trigger an "onHide" when clicked.
   */
  backdrop: PropTypes.oneOf(['static', true, false]),
  /**
   * Usually contains a Modal.Title, Modal.Body, and Modal.Footer
   * but can contain any content
   */
  children: PropTypes.any,
  /** Set whether the Escape key exits the modal */
  escapeExits: PropTypes.bool,
  /** Callback fired when the Modal transitions in */
  onEnter: PropTypes.func,
  /** Callback fired when the modal transitions out */
  onExit: PropTypes.func,
  /**
   * A callback fired when the header closeButton or non-static backdrop is
   * clicked. Required if either are specified.
   */
  onHide: PropTypes.func,
  /** When `true` The modal will show itself. */
  show: PropTypes.bool,
  /** Sets the size of the modal */
  size: PropTypes.oneOf(['small', 'medium', 'large'])
};

Modal.defaultProps = {
  animation: true,
  backdrop: true,
  escapeExits: true,
  onEnter: noop,
  onExit: noop,
  onHide: noop,
  show: false,
  size: 'medium',
  children: undefined
};

Modal.Header = Header;
Modal.Body = Body;
Modal.Footer = Footer;

export default Modal;
