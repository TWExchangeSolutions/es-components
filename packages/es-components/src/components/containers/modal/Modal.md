Use the `closeButton` attribute on the Modal.Header to show or hide the "X" button.

For accessbility, make sure to include the attribute `aria-haspopup='dialog'` on whatever link you create to activate the modal. If you need
an informational dialog that contains no focusable elements, use a popover or something else more appropriate.

```
<div>
  <Button aria-haspopup='dialog' handleOnClick={() => setState({show: true, size: 'small', hideCloseButton: true})} style={{marginRight:'15px'}}>Open Small Modal</Button>

  <Button aria-haspopup='dialog' handleOnClick={() => setState({show: true, size: 'medium', hideCloseButton: false})} style={{marginRight:'15px'}}>Open Medium Modal</Button>

  <Button aria-haspopup='dialog' handleOnClick={() => setState({show: true, size: 'large', hideCloseButton: false})}>Open Large Modal</Button>

  <Modal
    size={state.size}
    show={state.show}
    onHide={() => setState({show: false})}
  >
    <Modal.Header hideCloseButton={state.hideCloseButton}>This is the header.</Modal.Header>
    <Modal.Body>Body Content. Anim pariatur cliche reprehenderit, enim eiusmod high life accusamus terry richardson ad squid. 3 wolf moon officia aute, non cupidatat skateboard dolor brunch.This is the popover's content. Anim pariatur cliche reprehenderit, enim eiusmod high life accusamus terry richardson ad squid.</Modal.Body>
    <Modal.Footer>This is the footer. <Button handleOnClick={()=> setState({show: false})} style={{margin:0}}>Ok</Button></Modal.Footer>
  </Modal>
</div>
```
