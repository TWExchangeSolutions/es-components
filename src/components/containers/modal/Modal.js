import React from 'react';
import PropTypes from 'prop-types';
import styled, { ThemeProvider, withTheme } from 'styled-components';
import { noop } from 'lodash';
import BaseModal from 'react-overlays/lib/Modal';

import defaultTheme from '../../theme/defaultTheme';
import genId from '../../util/generateAlphaName';
import Fade from '../../util/Fade';
import Header from './ModalHeader';
import Body from './ModalBody';
import Footer from './ModalFooter';

const modalSize = {
  small: '300px',
  medium: '600px',
  large: '900px'
};

const DialogWrapper = styled(BaseModal)`
  height: 100%;
  left: 0;
  position: fixed;
  text-align: center;
  top: 0;
  width: 100%;
  z-index: 1040;
`;

const ModalDialogMedium = styled.div`
  cursor: default;
  display: inline-block;
  top: 2em;

  @media (min-width: ${props => props.theme.screenSize.tablet}) {
    margin: 30px auto;
    width: ${modalSize.medium};
  }
`;

const ModalDialogSmall = ModalDialogMedium.extend`
  @media (min-width: ${props => props.theme.screenSize.phone}) {
    margin: 30px auto;
    width: ${modalSize.small};
  }
`;

const ModalDialogLarge = ModalDialogMedium.extend`
  @media (min-width: ${props => props.theme.screenSize.desktop}) {
    width: ${modalSize.large};
  }
`;

const ModalContent = styled.div`
  background-clip: padding-box;
  background-color: ${props => props.theme.colors.white};
  box-shadow: 0 5px 15px rgba(0, 0, 0, 0.5);
  font-size: ${props => props.theme.sizes.baseFontSize}px;
  outline: 0;
  position: relative;
`;

class Modal extends React.Component {
  constructor(props, context) {
    super(props, context);

    this.state = {
      style: {},
      ariaId: genId()
    };
  }

  getChildContext() {
    return {
      modal: {
        onHide: this.props.onHide,
        ariaId: this.state.ariaId
      }
    };
  }

  render() {
    const {
      animation,
      backdrop,
      children,
      escapeExits,
      onEnter,
      onExit,
      onHide,
      show,
      size,
      theme
    } = this.props;

    const backdropStyle = {
      backgroundColor: theme.colors.black,
      bottom: 0,
      cursor: backdrop === 'static' ? 'auto' : 'pointer',
      left: 0,
      opacity: 0.5,
      position: 'fixed',
      right: 0,
      top: 0,
      zIndex: 'auto'
    };

    let ModalDialog;
    switch (size) {
      case 'small':
        ModalDialog = ModalDialogSmall;
        break;
      case 'large':
        ModalDialog = ModalDialogLarge;
        break;
      default:
        ModalDialog = ModalDialogMedium;
        break;
    }

    return (
      <ThemeProvider theme={theme}>
        <DialogWrapper
          backdrop={backdrop}
          backdropStyle={backdropStyle}
          keyboard={escapeExits}
          onEnter={onEnter}
          onExit={onExit}
          onHide={onHide}
          show={show}
          transition={animation ? Fade : undefined}
        >
          <ModalDialog
            size={size}
            aria-labelledby={this.state.ariaId}
            className="modal-dialog"
          >
            <ModalContent>{children}</ModalContent>
          </ModalDialog>
        </DialogWrapper>
      </ThemeProvider>
    );
  }
}

Modal.propTypes = {
  /** Open and close the Modal with a fade animation. */
  animation: PropTypes.bool,
  /**
   * Include a backdrop component. Specify 'static' for a backdrop that doesn't
   * trigger an "onHide" when clicked.
   */
  backdrop: PropTypes.oneOf(['static', true, false]),
  /** Sets whether clicking outside exits the modal */
  backdropClickExits: PropTypes.bool,
  /**
   * Usually contains a Modal.Title, Modal.Body, and Modal.Footer
   * but can contain any content
   */
  children: PropTypes.any,
  /** Set whether the Escape key exits the modal */
  escapeExits: PropTypes.bool,
  /** Callback fired when the Modal transitions in */
  onEnter: PropTypes.func,
  /** Callback fired when the modal transitions out */
  onExit: PropTypes.func,
  /**
   * A callback fired when the header closeButton or non-static backdrop is
   * clicked. Required if either are specified.
   */
  onHide: PropTypes.func,
  /** When `true` The modal will show itself. */
  show: PropTypes.bool,
  /** Sets the size of the modal */
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  /**
   * Theme object used by the ThemeProvider,
   * automatically passed by any parent component using a ThemeProvider
   */
  theme: PropTypes.object
};

Modal.defaultProps = {
  animation: true,
  backdrop: true,
  escapeExits: true,
  onEnter: noop,
  onExit: noop,
  onHide: noop,
  show: false,
  size: 'medium'
};

Modal.childContextTypes = {
  modal: PropTypes.shape({
    onHide: PropTypes.func,
    ariaId: PropTypes.string
  })
};

Modal.Header = Header;
Modal.Body = Body;
Modal.Footer = Footer;

const ThemedModal = withTheme(Modal);

ThemedModal.defaultProps = { theme: defaultTheme };

export default ThemedModal;
