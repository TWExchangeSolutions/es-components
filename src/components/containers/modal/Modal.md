Use the `closeButton` attribute on the Modal.Header to show or hide the "X" button.

For accessbility, make sure to include the attribute `aria-haspopup='dialog'` on whatever link you create to activate the modal.

```
<div>
  <Button aria-haspopup='dialog' handleOnClick={() => setState({show: true, size: 'small', showCloseButton: false})} style={{marginRight:'15px'}}>Open Small Modal</Button>

  <Button aria-haspopup='dialog' handleOnClick={() => setState({show: true, size: 'medium', showCloseButton: true})} style={{marginRight:'15px'}}>Open Medium Modal</Button>

  <Button aria-haspopup='dialog' handleOnClick={() => setState({show: true, size: 'large', showCloseButton: true})}>Open Large Modal</Button>

  <Modal
    size={state.size}
    show={state.show}
    onHide={() => setState({show: false})}
  >
    <Modal.Header closeButton={state.showCloseButton}>This is the header.</Modal.Header>
    <Modal.Body>Body Content. Anim pariatur cliche reprehenderit, enim eiusmod high life accusamus terry richardson ad squid. 3 wolf moon officia aute, non cupidatat skateboard dolor brunch.This is the popover's content. Anim pariatur cliche reprehenderit, enim eiusmod high life accusamus terry richardson ad squid.</Modal.Body>
    <Modal.Footer>This is the footer.</Modal.Footer>
  </Modal>
</div>
```
