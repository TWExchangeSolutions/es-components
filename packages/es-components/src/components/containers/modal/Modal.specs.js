/* eslint-env jest */

import React from 'react';
import { mountWithTheme } from 'styled-enzyme';
import Modal from './Modal';

describe('modal component', () => {
  let instance;
  const onHide = jest.fn();

  beforeEach(() => {
    instance = mountWithTheme(
      <Modal show onHide={onHide} animation={false}>
        <div className="body">Modal body text.</div>
      </Modal>
    );
  });

  it('mounts as expected', () => {
    expect(instance.find(Modal).length).toBe(1);
    expect(instance.find(Modal).prop('show')).toBe(true);
  });
});
