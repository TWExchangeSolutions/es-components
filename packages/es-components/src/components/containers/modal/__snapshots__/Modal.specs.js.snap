// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders different modal sections 1`] = `
<h4
  class="sc-bxivhb eouofC"
  id="abcdef"
>
  Header
</h4>
`;

exports[`renders different modal sections 2`] = `
<div
  class="sc-EHOje fLjynL"
>
  Body
</div>
`;

exports[`renders different modal sections 3`] = `
<div
  class="sc-bZQynM jggbBG"
>
  Footer
</div>
`;
